### DISCUSSION TITLE
com.openai.unity 7.3.2

### DISCUSSION THREAD

**Discussion by StephenHodgson** [RageAgainstThePixel/com.openai.unity#166] [Announcements]
## What's Changed
- Fixed parameter name in Threads.CreateMessageRequest
- Added Stream overload to Threads.FileUploadRequest

**Full Changelog**: https://github.com/RageAgainstThePixel/com.openai.unity/compare/7.3.1...7.3.2

<hr /><em>This discussion was created from the release <a href='https://github.com/RageAgainstThePixel/com.openai.unity/releases/tag/7.3.2'>com.openai.unity 7.3.2</a>.</em>

**Comment by weto4ka** [comment]
hey! impressive work :) 
Does your package support Webgl? or its only for Windows, Mac and Linux?

**Reply by StephenHodgson** [reply]
Yes it does 👍

**Reply by weto4ka** [reply]
Really cool stuff you've got with the OpenAI Azure/Unity package! I've been trying to get it to work for a WebGL build, but I'm running into some issues.

I'm currently on Unity 2023.3.0b3 and also tested it on the latest alpha (2023.3.0a18). Every time I try to build for WebGL, it just crashes on me. 😕

I thought it might be some conflict with other packages in my project, but nope, I got the same error even in a clean Unity 3D project. I did the usual – turned off compression, switched the API compatibility to .NET Framework, brought in your package and the demo scene. But when I hit build for WebGL, it's a no-go.

I'm not sure if I'm missing a step or need to tweak a setting somewhere. Do you have any ideas what might be causing the crash, or is there something else I should be doing?

I've attached the error message from the build. Any help would be super appreciated!

Cheers

Here is the build error I get:

![image](https://github.com/RageAgainstThePixel/com.openai.unity/assets/33402888/66d566cc-e373-4ee6-aaea-fed2a03a13ca)

```
Build completed with a result of 'Failed' in 33 seconds (33013 ms)
C:\Program Files\Unity\Hub\Editor\2023.3.0b3\Editor\Data\il2cpp\build\deploy\il2cpp.exe --convert-to-cpp --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/Assembly-CSharp.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/Mono.Security.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/mscorlib.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/Newtonsoft.Json.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/OpenAI.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/Sirenix.OdinInspector.Attributes.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/Sirenix.Serialization.Config.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/Sirenix.Serialization.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/Sirenix.Utilities.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/System.Configuration.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/System.Core.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/System.Data.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/System.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/System.Numerics.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/System.Runtime.Serialization.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/System.Xml.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/System.Xml.Linq.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/Unity.TextMeshPro.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/UnityEngine.AIModule.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/UnityEngine.AnimationModule.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/UnityEngine.AudioModule.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/UnityEngine.CoreModule.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/UnityEngine.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/UnityEngine.GridModule.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/UnityEngine.HierarchyCoreModule.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/UnityEngine.ImageConversionModule.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/UnityEngine.IMGUIModule.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/UnityEngine.InputForUIModule.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/UnityEngine.InputLegacyModule.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/UnityEngine.JSONSerializeModule.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/UnityEngine.Physics2DModule.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/UnityEngine.PhysicsModule.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/UnityEngine.PropertiesModule.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/UnityEngine.SharedInternalsModule.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/UnityEngine.SpriteShapeModule.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/UnityEngine.TextCoreFontEngineModule.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/UnityEngine.TextCoreTextEngineModule.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/UnityEngine.TextRenderingModule.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/UnityEngine.TilemapModule.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/UnityEngine.UI.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/UnityEngine.UIElementsModule.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/UnityEngine.UIModule.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/UnityEngine.UnityWebRequestAssetBundleModule.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/UnityEngine.UnityWebRequestAudioModule.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/UnityEngine.UnityWebRequestModule.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/UnityEngine.UnityWebRequestTextureModule.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/Utilities.Async.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/Utilities.Audio.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/Utilities.Encoding.Wav.dll --assembly=Library/Bee/artifacts/WebGL/ManagedStripped/Utilities.Rest.dll --generatedcppdir=C:/Users/Me/Test Azure OpenAI/Library/Bee/artifacts/WebGL/il2cppOutput/cpp --symbols-folder=C:/Users/Me/Test Azure OpenAI/Library/Bee/artifacts/WebGL/il2cppOutput/cpp/Symbols --enable-analytics --emit-method-map --dotnetprofile=unityaot-linux --usymtool-path=C:/Program Files/Unity/Hub/Editor/2023.3.0b3/Editor/Data\Tools\usymtool.exe --profiler-report --profiler-output-file=C:/Users/Me/Test Azure OpenAI/Library/Bee/artifacts/il2cpp_conv_gb1y.traceevents --print-command-line --static-lib-il2-cpp --data-folder=C:/Users/Me/Test Azure OpenAI/Library/Bee/artifacts/WebGL/il2cppOutput/data
Error: IL2CPP error for method 'System.Void System.Threading.<>_GenericWrapper::.ctor(System.Action`1<T>)' in assembly 'C:\Users\Me\Test Azure OpenAI\Library\Bee\artifacts\WebGL\ManagedStripped\mscorlib.dll'
System.InvalidOperationException: Attempting to return from void method 'System.Void System.Threading.<>_GenericWrapper::.ctor(System.Action`1<T>)' when there are values on the stack. Is this invalid IL code?
   at Unity.IL2CPP.MethodBodyWriter.WriteReturnStatement() + 0x4ea
   at Unity.IL2CPP.MethodBodyWriter.ProcessInstruction(Node, InstructionBlock, ResolvedInstruction) + 0x85e
   at Unity.IL2CPP.MethodBodyWriter.GenerateCodeRecursive(Node, ReadOnlyDictionary`2) + 0x827
   at Unity.IL2CPP.MethodBodyWriter.Generate() + 0x72f
   at Unity.IL2CPP.CodeWriters.CodeWriterExtensions.WriteMethodWithMetadataInitialization(IGeneratedMethodCodeWriter, String, Action`2, String, MethodReference, WritingMethodFor) + 0x2ef
   at Unity.IL2CPP.MethodWriter.WriteMethodDefinition(AssemblyWriteContext, IGeneratedMethodCodeWriter, MethodReference) + 0x40a
   at Unity.IL2CPP.SourceWriter.WriteTypesMethods(SourceWritingContext, IGeneratedMethodCodeWriter, TypeWritingInformation&, NPath, Boolean) + 0x560
   at Unity.IL2CPP.SourceWriters.SourceWriterBase`2.WriteItem(StreamWorkItemData`2) + 0x1fc
   at Unity.IL2CPP.Contexts.Scheduling.Streams.FileLevelParallelStreamManager`3.WorkerWriteItemsToFile(WorkItemData`2 data) + 0x1ed
   at Unity.IL2CPP.Contexts.Scheduling.PhaseWorkScheduler`1.WorkerLoop(Object data) + 0x60e

UnityEditor.GenericMenu:CatchMenu (object,string[],int)
```

**Reply by StephenHodgson** [reply]
@weto4ka I'm sorry to hear that.

I'll do a quick build in the project to make sure everything is good.

The exception doesn't really give us much info to go on, except that there was a problem converting some IL code.

**Reply by StephenHodgson** [reply]
@weto4ka I was able to build and ran my sample chat scene in webgl just fine. 

Unity 2022.3 LTS

**Reply by weto4ka** [reply]
Thanks for the quick response! I haven't tried it on Unity 2022.3 LTS yet – our project is currently on the latest Beta version (2023.3.0b3).

Would it be possible for you to take a look at why it might not be working on Unity 2023.3.0b3, or could you point me in the right direction on what to check? Any guidance would be super helpful!

Appreciate your help!

**Reply by StephenHodgson** [reply]
Just did a build on Unity 2023.3.0b3, and it indeed failed. 

Sorry, but I don't support the beta builds by default.
I can look into it and making sure this isn't a blocker once 2023.3 comes out of beta, but this seems like a bug with Unity IL2CPP generation.

**Reply by weto4ka** [reply]
sure, I understand. thank you)

**Reply by StephenHodgson** [reply]
https://forum.unity.com/threads/cant-build-to-webgl-il2cpp-error.1540694/

**Reply by weto4ka** [reply]
thanks, yes, I reported it to Unity) Hopefully they fix it :)
My previous bug report for them is still unanswered, they just disappeared at some point and stopped replying 😅.

**Reply by StephenHodgson** [reply]
Well there is a new case number in thread provided above. Seems other people were able to create repo cases. Guess we will see.

**Reply by scott-ferguson-unity** [reply]
@StephenHodgson Your issue is at 
https://github.com/RageAgainstThePixel/com.utilities.async/blob/138003d8f5d931f3d57ff737931a7669f31b9ae4/Utilities.Async/Packages/com.utilities.async/Editor/WebGL/WebGLPostBuildCallback.cs#L165.  

That generates call to `static void System.Object::.ctor()` which doesn't exist.  You need to make your `MethodReference` to an instance method`

`ilProcessor.Emit(OpCodes.Call, new MethodReference(".ctor", moduleDefinition.TypeSystem.Void, moduleDefinition.TypeSystem.Object) { HasThis = true });`

**Reply by StephenHodgson** [reply]
Thank you for taking a look @scott-ferguson-unity!

Was this removed? It was working previously.
So just needs to changed for 2023.+?

**Reply by scott-ferguson-unity** [reply]
It needs to be changed for all versions, `System.Object` has never had a static constructor.  Even if it had the call at 165 is to a static method, so the push at [164](https://github.com/RageAgainstThePixel/com.utilities.async/blob/138003d8f5d931f3d57ff737931a7669f31b9ae4/Utilities.Async/Packages/com.utilities.async/Editor/WebGL/WebGLPostBuildCallback.cs#L164) unbalances the IL stack, which would have always been a problem.  I have no idea why this would have ever worked.

**Reply by StephenHodgson** [reply]
Good to know! 😅 

I'll make the changes. Thanks!

**Reply by StephenHodgson** [reply]
Fixed in https://github.com/RageAgainstThePixel/com.utilities.async/pull/13
